README.md

```diff
@@ -108,24 +108,43 @@ which is explained in detail below.
-   This is necessary because GitHub requires releases to be tagged and the following workflow uploads a release.
-3. Run the [workflow](./.github/workflows/publish-framework.yml) for publishing a new version of the binary framework.
-   Use the tag you just created as "branch" to run from and input `<version>` (i.e. without the counter for "Version").
-4. Run `make swift-bindings` locally to regenerate the Swift bridge sources.
-5. Update `Package.swift` with the updated `url` and `checksum` of the binary framework.
-   The workflow prints the checksum as the last step of its execution.
-6. Commit the changes to `Sources/ConcordiumWalletCrypto/crypto.swift` and `Package.swift`
-   (no other files should have changes)
-   and push an annotated tag named by the version for the new commit:
+Due to the dual nature of this repository, the release process is a little cumbersome,
+so please follow the instructions below carefully.
+
+The entire process should be done as a single PR (with no unrelated changes);
+see [#11](https://github.com/Concordium/concordium-wallet-crypto-swift/pull/11) for a template.
+To preserve tags, the PR must be merged without squashing.
+
+Steps for building and releasing a new version `<version>` of the package:
+
+1. Create and checkout a release branch named `release/<version>`.
+2. Bump the version to `<version>` in `Cargo.toml` and
+   [insert the changelog header](https://github.com/Concordium/concordium-wallet-crypto-swift/pull/11/files#diff-06572a96a58dc510037d5efa622f9bec8519bc1beab13c9f251e97e657a9d4ed).
+   Commit and push the change to the release branch.
+3. Create and push a (regular) tag named `build/<version>-<build-version>` for the new commit,
+   where `<build-version>` starts out at `0`.
+   Tagging is necessary because the workflow to be run in the next step uploads a release,
+   and GitHub requires releases to be tagged.
+4. Run the [workflow](./.github/workflows/publish-framework.yml) for publishing a new version of the binary framework.
+   Use the tag you just created as "branch" to run from and input `<version>` (i.e. without the build number).
+   If the build fails, commit a fix to the branch and return to step 3 with `<build-version>` incremented by 1.
+5. Run `make swift-bindings` locally to regenerate the Swift bridge sources.
+6. Update `Package.swift` with the updated `url` and `checksum` of the binary framework.
+   The checksum is part of the release built by the workflow above as the file `CHECKSUM`.
+7. Commit the changes to `Sources/ConcordiumWalletCrypto/generated.swift` and `Package.swift`
+   (no other files should have changes).
+   IMPORTANT: The generated Swift file is ignored by Git to prevent it from being updated untimely,
+   so it must be force added before committing:
+   ```shell
+   git add -f ./Sources
+   ```
+8. Merge PR for release branch *without squashing*.
+9. On the resulting *merge commit* into `main`, push an *annotated* tag named by the version:
```

Comment:
This whole thing sounds a little daunting tbh. For other repositories we have success following a process like this:

- create PR with version bump -> merge to main
- tag merge commit on main with `release/<package>/*.*.*` -> trigger github release flow
- release must be approved by other member of corresponding github group
- releases according to what is being published

It seems like here there are a lot of steps required to automate things. Would it be possible to follow a similar process here? This might not be the PR to introduce these changes, but I could get a little worried that this will be hard for others to follow and it would be better to streamline the process.

Reply:
It is a bit special which is also why the instructions are so prescriptive. As explained, the reason it that this repo really contains two projects, a Rust crate and a Swift package.

You have to release the framework and Swift package with separate workflows. Otherwise the workflow would have to also commit changes that are hard to do correctly automatically.

The way to simplify it would be to split the repo and thus also the release process. That could make sense, especially if `concordium-wallet-crypto-uniffi` (the Rust crate) were to be used for other purposes (i.e. SDKs in other languages). But I think the current situation is fine because it's so thoroughly documented.

Note that from a security perspective, everything related to releases is completely confined within GitHub. So the workflows don't refer to any external secrets. Which also means that anyone in the org could publish a bad framework, even without using the workflow. I don't know if GitHub allows limiting access to this kind of thing.